### PR TITLE
Antonio - UI Design Patterns Challenge

### PR DESCRIPTION
**Pull request by aamayorga**

### CODE REVIEW

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 1]
```diff
@@ -5,6 +5,8 @@
 import UIKit
 
 public final class FeedViewController: UITableViewController, UITableViewDataSourcePrefetching {
+	@IBOutlet weak var errorView: ErrorView!
```
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:8]: Hi @aamayorga! The `FeedViewController` is wired correctly with the `ErrorView` outlet in the storyboard.

However, the `ErrorView` is not wired with the `button` outlet in the storyboard. That's why it's crashing.

You need to connect the `ErrorView` with the `button` outlet in the storyboard contained in the MVVM folder. You can follow the MVC one as a reference ✅

**Comment by caiozullo** [comment]
Hi @aamayorga! 

The build is failing. Please, have a look and update the PR name following the instructions in the readme file.

Let me know when it's ready to review. Thank you! 🙂

**Comment by aamayorga** [comment]
@caiozullo The build is working now, hope everything else checks out 🙏

File: MVC Module/Source/Feed Presentation/es.lproj/Feed.strings [hunk 2]
```diff
@@ -0,0 +1,2 @@
+"FEED_VIEW_TITLE" = "Mi Feed";
+"FEED_VIEW_CONNECTION_ERROR" = "Un mensaje de error";
```
Comment (caiozullo) [inline MVC Module/Source/Feed Presentation/es.lproj/Feed.strings:2]: The error message should describe the actual error. See the other ones in the MVC template provided as a reference.

File: MVP Module/Source/Feed Presentation/en.lproj/Feed.strings [hunk 3]
```diff
@@ -1,2 +1,3 @@
 
 "FEED_VIEW_TITLE" = "My Feed";
+"FEED_VIEW_CONNECTION_ERROR" = "An error message";
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/en.lproj/Feed.strings:3]: The error message should describe the actual error. See the other ones in the MVC template provided as a reference.

File: MVP Module/Source/Feed Presentation/es.lproj/Feed.strings [hunk 4]
```diff
@@ -0,0 +1,2 @@
+"FEED_VIEW_TITLE" = "Mi Feed";
+"FEED_VIEW_CONNECTION_ERROR" = "Un mensaje de error";
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/es.lproj/Feed.strings:2]: The error message should describe the actual error. See the other ones in the MVC template provided as a reference.

File: MVP Module/Source/Feed Presentation/pt-BR.lproj/Feed.strings [hunk 5]
```diff
@@ -1,2 +1,2 @@
-
-"FEED_VIEW_TITLE" = "Meu Feed";
+"FEED_VIEW_TITLE" = "Mi Feed";
+"FEED_VIEW_CONNECTION_ERROR" = "Un mensaje de error";
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/pt-BR.lproj/Feed.strings:2]: The error message should describe the actual error. See the other ones in the MVC template provided as a reference.

File: MVP Module/Source/Feed Presentation/FeedPresenter.swift [hunk 6]
```diff
@@ -5,6 +5,10 @@
 import Foundation
 import FeedFeature
 
+protocol FeedErrorView {
+	func show(_ viewModel: FeedErrorViewModel)
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/FeedPresenter.swift:9]: The method could be called `display` to match the other implementations. Consistency is important in the codebase.

File: MVP Module/Source/Feed Presentation/FeedPresenter.swift [hunk 7]
```diff
@@ -16,10 +20,12 @@ protocol FeedView {
...
+	private let errorView: FeedErrorView
 	
-	init(feedView: FeedView, loadingView: FeedLoadingView) {
+	init(feedView: FeedView, loadingView: FeedLoadingView, errorView: FeedErrorView) {
 		self.feedView = feedView
 		self.loadingView = loadingView
+		self.errorView = errorView
 	}
 	
 	func didStartLoadingFeed() {
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/FeedPresenter.swift:31]: The Presenter decides when to show the error message, so it should also decide when to hide the message.

This way, the ViewController doesn't need to decide when to hide the error. It's easier to manage state when all changes are centralized in a single component.

For example, in the `didStartLoadingFeed()` method, the Presenter can call  `errorView.display(FeedErrorViewModel(errorMessage: .none))` to hide the error.

File: MVP Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 8]
```diff
@@ -33,6 +36,10 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 		}
 	}
 	
+	func show(_ viewModel: FeedErrorViewModel) {
+		errorView.isHidden = false
```
Comment (caiozullo) [inline MVP Module/Source/Feed UI/Controllers/FeedViewController.swift:40]: The implementation should check if the viewModel contains an error message.

If it does, show the error.
Else, hide the error.

And it should use the provided methods `errorView.show(message: message)`  and `errorView.hideMessage()`.

File: MVVM Module/Source/Feed Presentation/el.lproj/Feed.strings [hunk 9]
```diff
@@ -1,2 +1,3 @@
 
 "FEED_VIEW_TITLE" = "Το Feed μου";
+"FEED_VIEW_CONNECTION_ERROR" = "éna mínyma sfálmatos";
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/el.lproj/Feed.strings:3]: The error message should describe the actual error. See the other ones in the MVC template provided as a reference.

File: MVVM Module/Source/Feed Presentation/en.lproj/Feed.strings [hunk 10]
```diff
@@ -1,2 +1,3 @@
 
 "FEED_VIEW_TITLE" = "My Feed";
+"FEED_VIEW_CONNECTION_ERROR" = "An error message";
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/en.lproj/Feed.strings:3]: The error message should describe the actual error. See the other ones in the MVC template provided as a reference.

File: MVVM Module/Source/Feed Presentation/es.lproj/Feed.strings [hunk 11]
```diff
@@ -0,0 +1,3 @@
+
+"FEED_VIEW_TITLE" = "Mi Feed";
+"FEED_VIEW_CONNECTION_ERROR" = "Un mensaje de error";
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/es.lproj/Feed.strings:3]: The error message should describe the actual error. See the other ones in the MVC template provided as a reference.

File: MVVM Module/Source/Feed Presentation/pt-BR.lproj/Feed.strings [hunk 12]
```diff
@@ -1,2 +1,3 @@
 
 "FEED_VIEW_TITLE" = "Meu Feed";
+"FEED_VIEW_CONNECTION_ERROR" = "Uma mensagem de erro";
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/pt-BR.lproj/Feed.strings:3]: The error message should describe the actual error. See the other ones in the MVC template provided as a reference.

File: MVP Module/Source/Feed Presentation/FeedPresenter.swift [hunk 13]
```diff
@@ -33,5 +39,6 @@ final class FeedPresenter {
 	
 	func didFinishLoadingFeed(with error: Error) {
 		loadingView.display(FeedLoadingViewModel(isLoading: false))
+		errorView.show(FeedErrorViewModel(errorMessage: error.localizedDescription))
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/FeedPresenter.swift:42]: The error should be localized with `Localized.Feed.loadError` (like in the MVC and MVVM solution).

**Review by caiozullo — COMMENTED** [review]
Hi @aamayorga, 

You almost passed the challenge!

The `es` localization is not working. It's missing the`es` resources in the main app target. When adding a new localization, make sure to select all resource files - including storyboards. If you check the project settings, only 3 resources are localized (but should be 4 like the other localizations!).

Moreover, the project is missing integration tests for the error messages. You can check how other students did it in the other PRs.

I also left some review comments.

Please, let me know once you're done so I can review it again. Thank you! 🙂

**Comment by aamayorga** [comment]
Hi @caiozullo, I have made the changes you asked for and would like you to take another look. 😄

**Review by caiozullo — COMMENTED** [review]
Hi @aamayorga,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅
- Your code is clean and handles edge cases ✅
- The UI and animations look and behave as expected ✅
- Light and Dark mode support ✅
- Added an extra language ✅

Great job!
